compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1247,6 +1244,30 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 // the "projection predicate" for:
                 //
                 // `<T as Iterator>::Item = u32`
+                let assoc_item_def_id = projection_ty.skip_binder().item_def_id;
+                let def_kind = tcx.def_kind(assoc_item_def_id);
+                match (def_kind, term) {
+                    (hir::def::DefKind::AssocTy, ty::Term::Ty(_))
+                    | (hir::def::DefKind::AssocConst, ty::Term::Const(_)) => (),
+                    (_, _) => {
+                        let got = if let ty::Term::Ty(_) = term { "type" } else { "const" };
+                        let expected = if let hir::def::DefKind::AssocTy = def_kind {
+                            "associated const"
+                        } else if let hir::def::DefKind::AssocTy = def_kind {
+                            "associated type"
+                        } else {
+                            // This should be unreachable
+                            "[unknown]"
+                        };
+                        tcx.sess
+                            .struct_span_err(
+                                tcx.def_span(assoc_item_def_id),
```

Comment:
Keep the `binding.span` here, but use the def_span for the label or a secondary note.